CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,226 @@
+
+The API chosen here is for the minimum API necessary for dApp <-> Wallet interactions without convenience functions that don't strictly need the wallet's state to work. The API here is for now also only designed for Shelley's Mary hardfork and thus has NFT support. When Alonzo is released with Plutus support this API will have to be extended.
+
+### cardano.get_utxos(amount: cbor\<value> = undefined, paginate: Paginate = undefined): Promise\<TransactionUnspentOutput[] | undefined>
+
+Errors: `APIError`, `PaginateError`
+
+If `amount` is `undefined`, this shall return a list of all UTXOs (unspent transaction outputs) controlled by the wallet. If `amount` is not `undefined`, this request shall be limited to just the UTXOs that are required to reach the combined ADA/multiasset value target specified in `amount`, and if this cannot be attained, `undefined` shall be returned. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_balance(): Promise\<cbor\<value>>
+
+Errors: `APIError`
+
+Returns the total balance available of the wallet. This is the same as summing the results of `cardano.get_utxos()`, but it is both useful to dApps and likely already maintained by the implementing wallet in a more efficient manner so it has been included in the API as well.
+
+### cardano.get_used_addresses(paginate: Paginate = undefined): Promise\<\<cbor<address>[]>
+
+Errors: `APIError`
+
+Returns a list of all used (included in some on-chain transaction) addresses controlled by the wallet. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_unused_addresses(): Promise\<\<cbor<address>[]>
+
+Errors: `APIError`
+
+Returns a list of unused addresses controlled by the wallet.
+
+### cardano.get_change_address(): Promise\<\<cbor<address>>
+
+Errors: `APIError`
+
+Returns an address owned by the wallet that should be used as a change address to return leftover assets during transaction creation back to the connected wallet. This can be used as a generic receive address as well.
+
+### cardano.sign_tx(tx: cbor\<transaction_body>, metadata: cbor\<auxiliary_data> = undefined): Promise\<cbor\<transaction>>
+
+Errors: `APIError`, `TxSignError`
+
+Requests that a user sign the supplied transaction body. The wallet should ask the user for permission, and if given, try to sign the supplied body and return a signed transaction. If the wallet could not sign the transaction, `TxSignError` shall be returned with the `ProofGeneration` code. Likewise if the user declined it shall return the `UserDeclined` code.
+
+### cardano.sign_tx_input(tx: cbor\<transaction_body>, index: number): Promise\<cbor\<transaction_witness_set>>
```

Comment:
This method seems redundant with the previous one. I'd suggest replacing `index: number` with an optional `index: number[]` argument on the previous method; if empty or absent, all inputs should be witnessed, otherwise, only the specified indexes. This makes the API more orthogonal and easier to digest.

Reply:
I planned on removing the sign tx input entirely and instead having some kind of more general partial signing functionaltity as discussed earlier in this PR's comments with alessandro.